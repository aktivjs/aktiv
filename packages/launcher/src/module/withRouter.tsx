import React from 'react';
import { useLocation, useParams } from 'react-router-dom';
import { useRouter } from '../hooks';
import type { Params } from 'react-router-dom';
import type { Location } from 'react-router';
import type { UseRouterReturns } from '../hooks';
import type {
    ComponentType,
    PropsWithoutRef,
    ForwardRefExoticComponent,
    RefAttributes,
} from 'react';

export type HocExtraProps = {
    location: Location;
    params: Params;
    history: {
        go: typeof history.go;
        goBack: typeof history.back;
        goForward: typeof history.forward;
        location: Location;
        push: UseRouterReturns['redirect'];
        replace: UseRouterReturns['replace'];
    };
};

const globalHistory = window.history;

export type AktivHistory = {
    go(): void;
    go(delta?: number): void;
    goBack(): void;
    goForward(): void;
};

const aktivHistory: AktivHistory = {
    go(delta?: number) {
        return globalHistory.go(delta);
    },
    goBack() {
        return globalHistory.back();
    },
    goForward() {
        return globalHistory.forward();
    },
};

function withRouter<CProps, R>(
    Com: ComponentType<CProps & HocExtraProps>,
): ForwardRefExoticComponent<PropsWithoutRef<CProps> & RefAttributes<R>> {
    return React.forwardRef<R, CProps>((props, ref) => {
        const routerLocation = useLocation();
        const params = useParams();
        const router = useRouter();
        const wrapperHistory = {
            ...aktivHistory,
            push: router.redirect,
            replace: router.replace,
            location: routerLocation,
        };
        const extraProps = {
            params,
            location: routerLocation,
            history: wrapperHistory,
            ...props,
        };

        return <Com {...extraProps} ref={ref} />;
    });
}

export default withRouter;
