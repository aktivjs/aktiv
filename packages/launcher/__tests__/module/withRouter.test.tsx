import React, { Component, FC } from 'react';
import withRouter from '../../src/module/withRouter';
import Launcher from '../../src/module/app';
import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const spyPageRenderTime = jest.fn(() => {});
// eslint-disable-next-line @typescript-eslint/no-empty-function
const spyChildRenderTime = jest.fn(() => {});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
class DemoPage extends Component<any, any> {
    // eslint-disable-next-line no-useless-constructor
    constructor() {
        // eslint-disable-next-line prefer-rest-params,@typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line prefer-rest-params
        super(...Array.from(arguments));
    }

    handleClick = () => {
        // eslint-disable-next-line no-invalid-this
        this.props?.history.push('/child');
    };

    // eslint-disable-next-line class-methods-use-this
    render() {
        spyPageRenderTime();

        return (
            <>
                <h1>Test Router</h1>
                <button onClick={this.handleClick}>跳转</button>
            </>
        );
    }
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const WithRouterPage = withRouter(DemoPage);

const ChildComponent: FC = () => {
    spyChildRenderTime();

    return <h1>child</h1>;
};

describe('withRouter', () => {
    afterAll(() => {
        jest.clearAllMocks();
    });

    const homeRoute = {
        path: '/',
        component: WithRouterPage,
        // children: [
        //     {
        //         path: '/child',
        //         component: ChildComponent,
        //     },
        // ],
    };
    const app = new Launcher({
        hash: false,
        routes: [
            homeRoute,
            {
                path: '/child',
                component: ChildComponent,
            },
        ],
    });

    it('when render app, should can get router & query & match args', () => {
        act(() => {
            app.start();
        });
        expect(spyPageRenderTime).toHaveBeenCalledTimes(1);
        act(() => {
            userEvent.click(screen.getByText('跳转'));
        });
        expect(spyChildRenderTime).toHaveBeenCalledTimes(1);
    });
});
