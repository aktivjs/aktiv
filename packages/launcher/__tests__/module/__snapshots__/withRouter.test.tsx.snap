// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRouter when render app, should can get router & query & match args 1`] = `
<body>
  <div
    data-testid="root"
    id="root"
  >
    <h1>
      Test Router
    </h1>
    <button>
      跳转
    </button>
  </div>
</body>
`;

exports[`withRouter when render app, should can get router & query & match args 2`] = `
<body>
  <div
    data-testid="root"
    id="root"
  >
    <h1>
      Test Router
    </h1>
    <button>
      跳转
    </button>
  </div>
</body>
`;
